### ISSUE TITLE
feat: Support typesafe inputs in `t.middleware` created middlewares

### ISSUE THREAD

**Issue by Nick-Lucas** [trpc/trpc#4191]
### Describe the feature you'd like to request

This is based on #4190 

Currently we can create a typesafe input within a middleware like this:

![image](https://user-images.githubusercontent.com/8896153/232254100-ed261297-e8ef-448f-8e20-4b049e2aae53.png)

However the de-facto way to create a re-usable and composable middleware is with `t.middleware`, not `procedure.use()`. 

When using the more composable approach we give up the type-safety:

![image](https://user-images.githubusercontent.com/8896153/232254147-12bbc001-8c74-4989-bd5b-d7940f9b97f7.png)

It would be a good step for composition to support type-safe inputs via `t.middleware`

### Describe the solution you'd like to see

There are a couple possibilities:

#### TypeScript interface

Allow `t.middleware` to take a generic parameter defining a mandatory input interface which must be met when it's used:

```ts
const orgMiddleware = t.middleware<{ orgId: number }>((opts) => {
      if (checkOrg(opts.input.orgId)) {
        // do checks and throw if needed

        return opts.next();
      }

      throw new TRPCError({
        code: 'FORBIDDEN',
      });
})

const scopedProcedure = publicProcedure
  .input(z.object({ orgId: z.number() }))
  .use(orgMiddleware)

const scopedProcedure = authProcedure
  .input(z.object({ orgId: z.number() }))
  .use(orgMiddleware
```

#### Zod 'preprocess' style 

```ts
const orgMiddleware = t.middleware<{ orgId: number }>((opts) => {
      if (checkOrg(opts.input.orgId)) {
        // do checks and throw if needed

        return opts.next();
      }

      throw new TRPCError({
        code: 'FORBIDDEN',
      });
  },
  // we pass a second property for the input parser, it will stack and merge just like .input does over multiple calls
  z.object({ orgId: z.number() })
)

const scopedProcedure = publicProcedure
  .use(orgMiddleware)

const scopedProcedure = authProcedure
  .use(orgMiddleware
```

### Describe alternate solutions

Alternatively we simply document a workaround for the issue, for instance this will provide the type safety to the client, and to the internals but is more verbose and doesn't enforce the use of the input :

```ts
const OrgObj = z.object({ orgId: z.number() })

const orgMiddleware = t.middleware<{ orgId: number }>((opts) => {
      const { orgId } = OrgObj.parse(opts.input)

      if (checkOrg(orgId)) {
        // do checks and throw if needed

        return opts.next();
      }

      throw new TRPCError({
        code: 'FORBIDDEN',
      });
})

const scopedProcedure = publicProcedure
  .input(OrgObj)
  .use(orgMiddleware)

const scopedProcedure = authProcedure
  .input(OrgObj)
  .use(orgMiddleware

// This will very likely blow up at runtime and not provide the type safety we generally go for!
const scopedProcedure = otherProcedure
  .use(orgMiddleware
```

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

**Comment by juliusmarminge** [comment]
Theres a third approach but not sure how much refactoring that would take to support - let middlewares alter inputs down the chain:

```

const orgMiddleware = t.middleware((opts) => {
      const input = OrgObj.parse(opts.input)

      if (checkOrg(input.orgId)) {
        // do checks and throw if needed

        return opts.next({
          input,
          // ctx: { ... },
        });
      }

      throw new TRPCError({
        code: 'FORBIDDEN',
      });
})
```

**Comment by Nick-Lucas** [comment]
> Theres a third approach but not sure how much refactoring that would take to support - let middlewares alter inputs down the chain:> 

I like this a lot, it's more canonical to how middlewares work thanks to `ctx`

**Comment by KATT** [comment]
Yesss, let's do this.

The one that @juliusmarminge points out has one important caveat - parsing here will throw an INTERNAL_SERVER_ERROR unless you do it in a safe way, so I wouldn't say it's more intuitive b/c of that reason. But allowing input swapping the same way we do for context would be nice anyway, so I'd like that too if it's even possible, I think I tried once but has improved since too.

My question on @Nick-Lucas' proposal is if we will want more generics passed as well, what if you want to also define `output` for whatever reason?

Maybe it should be:

```ts
const rateLimiter = t.middleware<{ input: { orgId: number }}>((opts) => {
  // ...
})
```

**Comment by Nick-Lucas** [comment]
> Yesss, let's do this.
> 
> The one that @juliusmarminge points out has one important caveat - parsing here will throw an INTERNAL_SERVER_ERROR unless you do it in a safe way, so I wouldn't say it's more intuitive b/c of that reason. But allowing input swapping the same way we do for context would be nice anyway, so I'd like that too if it's even possible, I think I tried once but has improved since too.
> 
> My question on @Nick-Lucas' proposal is if we will want more generics passed as well, what if you want to also define `output` for whatever reason?
> 
> Maybe it should be:
> 
> ```ts
> const rateLimiter = t.middleware<{ input: { orgId: number }}>((opts) => {
>   // ...
> })
> ```

Bit of a shower thought on that note, but I wonder if there's an opportunity in designing the solution to this, to better support 3rd party middlewares?

If a middleware can statically define its required input, required context, required meta, and be exported from some 3rd party library, then there's no need to couple a middleware to the initTRPC of the consuming application and vendors can provide 1-line integrations?

I see this as a current problem for adoption/ecosystem, as most middleware systems at least appear easier to put into a library for Auth tools and such?

**Comment by KATT** [comment]
> Bit of a shower thought on that note, but I wonder if there's an opportunity in designing the solution to this, to better support 3rd party middlewares?
> 
> If a middleware can statically define its required input, required context, required meta, and be exported from some 3rd party library, then there's no need to couple a middleware to the initTRPC of the consuming application and vendors can provide 1-line integrations?
> 
> I see this as a current problem for adoption/ecosystem, as most middleware systems at least appear easier to put into a library for Auth tools and such?

Agreed. It would be great if one could define a middleware _outside_ of the "`t`-variable" where you say that the middleware should satisfy at minimum a context containing X, an input Y, etc
